### PR TITLE
Determine the first date to show feedback question banner more randomly

### PR DESCRIPTION
**Pull request by sgiehl**
### Description:

This should randomize the first time the feedback banner is shown

@tsteur is this random enough?

follow-up to #18308

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/Feedback/Feedback.php [hunk 1]
```diff
@@ -123,15 +123,19 @@ public function showQuestionBanner()
...
-
-        // if is new user or old user field not exist
-        if ($nextReminderDate === false) {
-            //new user extend to 6 month, don't show banner
-            $nextReminder = Date::now()->getStartOfDay()->addMonth(6)->toString('Y-m-d');
+        // If there isn't any reminder date set, or never remind me was selected previously (-1) we determine a new date
+        if ($nextReminderDate === false || $nextReminderDate <= 0) {
+
+            // if user was created within the last 6 months, we set the date to 6 months after his creation date
+            $userCreatedDate = Piwik::getCurrentUserCreationData();
```
Comment (tsteur) [inline plugins/Feedback/Feedback.php]: There's a typo in the method name, it should be `date` probably ? Be good to change it in this release before it ends up being used somewhere else

**Review by tsteur — CHANGES_REQUESTED** [review]
@sgiehl it looks random. However, if I see this correct there is one problem:

> Date::factory($userCreatedDate)->addMonth(6)->getTimestamp() > $now

This would apply to a lot of users after updating? Would potentially most users get to see this after updating if their users was created more than 6 months ago? I think I mention something like we may want to check something like if the user was created between 6 and 6.5 months ago. 

For this to work we might actually need to do a modulo as a user might have been created 1 year ago. Would we need to find out number of days since signing up, and then do a modulo (like $numDaysSinceRegister modulo 6*30)?

File: plugins/Feedback/Feedback.php (hunk 1 above)
Comment (sgiehl) [reply plugins/Feedback/Feedback.php]: fixed that

**Comment by sgiehl** [comment]
@tsteur updated the code. I had actually used the wrong line of code. It should now do what I had described in the PR.
Users created within the last 6 months will see the banner 6 months after the creation. Users that are older will get a random date within the next 6 months.

**Comment by tsteur** [comment]
@sgiehl I'm not sure if it would still cause every user that was created like a year ago to see this at the same time?

Like assuming creation date was 1.5 years ago. That means `!empty($userCreatedDate) && Date::factory($userCreatedDate)->addMonth(6)->getTimestamp() > $now === true` meaning it goes into the if block.

It then adds 6 months to the creation date in `$nextReminder = Date::factory($userCreatedDate)->addMonth(6)->toString('Y-m-d');`. Meaning nextReminder is in this case 1 year ago. Meaning next page load we would show it?

The same would happen for users signed up 1 year ago and 2 years ago etc. Basically, as soon as they upgrade to the next version, they would all get to see the feedback screen?

Or maybe I'm missing something?

**Comment by sgiehl** [comment]
@tsteur it actually should only go into the if block if the creation date is within the past 6 months. If it is older `(creationdate + 6 months)` should be lower than `$now`. Or did I get something wrong in the if statement :thinking:

**Comment by justinvelluppillai** [comment]
Looks right to me, if the creation date plus six months is less than now it will skip the if block

**Comment by tsteur** [comment]
I see, makes sense. Sorry about that. So in that case we then just pick something randomly. I thought initially to make it then continuous so we show it after 1 year, 1.5 years, 2 years but we make it instead randomly. That's why I read it wrong. That should be all fine though and works too 👍 

Ideally there were tests for it @sgiehl as working with date etc there's always quickly a bug introduced. But I'm happy to skip it to get it into the RC.

**Comment by tsteur** [comment]
Meaning feel free to merge if tests pass.

**Comment by sgiehl** [comment]
@tsteur it should only be the first date that is picked randomly in that case. Once it was shown and clicked away it will be shown again after 6 months.

**Comment by tsteur** [comment]
Yep 👍  all good. 

I meant the first one we show after 1 year, or 1.5 year etc. But that's not important actually just something I had in my mind somehow. Picking the first date randomly is probably even better 👍  so all good

**Pull request merged**